### DISCUSSION TITLE
Intellij IDEA does not support Mill-based Scala 3 projects

### DISCUSSION THREAD

**Discussion by informarte** [com-lihaoyi/mill#1455] [Development]
I upgraded a project to Scala 3, fixed all issues, and now I have a green build.

As next step, I wanted to test the Scala 3 support of IntelliJ IDEA (latest version 2021.2.1). So I regenerated the project files, started the IDE, and changed some if-then-else statements to the new Scala 3 syntax. However, the IDE did not like the change: The changed parts were marked as syntactically wrong and the code did not compile in the IDE, although it compiled on command line. So it seems that the IDE is not aware that the project uses Scala 3.

Now I wonder what is wrong? Is there a secret switch in the IDE? Are the project files lacking something?

I am using Mill 0.9.9.

**Comment by informarte** [comment]
To track down the issue, I created a minimal sbt project:
```
name := "scala3-sbt-idea-test"
version := "0.1"
scalaVersion := "3.0.2"
```
After importing this project, IDEA is aware of Scala 3 .

Next I created a minimal Mill project:
```
object scala3MillIdeaTest extends ScalaModule {
  def scalaVersion = "3.0.2"
}
```
After creating the IDEA project (using `./mill mill.scalalib.GenIdea/idea`) and importing it, IDEA is NOT aware of Scala 3.

I glanced over and compared the IDEA project files but could not find anything pointing to Scala 3 except for the dependency on the Scala 3 library which is present in both IDEA projects (in addition to the Scala 2 library).

**Reply by lefou** [reply]
After fiddling around with IDEA, there is actual a difference in file `.idea/libraries/scala3_library_3_3_0_2_jar.xml` or me:

```diff
 <component name="libraryTable">
     <library name="scala3-library_3-3.0.2.jar" type="Scala">
         <properties>
+           <language-level>Scala_3_0</language-level>
             <compiler-classpath>
```

**Reply by informarte** [reply]
I see this difference, too: In the project files generated by `./mill mill.scalalib.GenIdea/idea`, there is no language level.

(In the sbt project, the file path is `.idea/libraries/sbt__org_scala_lang_scala3_library_3_3_0_2_jar.xml`.)

**Reply by informarte** [reply]
Meanwhile, I had filed an issue with the IDEA bug tracker: https://youtrack.jetbrains.com/issue/SCL-19514

It turned out that importing Mill-based Scala 3 projects via BSP works in the sense that IDEA can compile Scala 3 code and run tests. (Other things like error highlighting and completion do not yet work so well, unfortunately, but these issues seem to be unrelated to the import.)

As the IntelliJ support is considered legacy (https://com-lihaoyi.github.io/mill/mill/Intro_to_Mill.html#_intellij_support_legacy), I conclude that BSP is the way to go, right?

**Reply by lefou** [reply]
We have currently noone who maintains BSP support. I've create a project to collect all issues related to BSP: https://github.com/com-lihaoyi/mill/projects/1
I try my best to keep IntelliJ IDEA support usable.

**Reply by lefou** [reply]
I opened a PR to fix the missing language-level property: #1486 . Any feedback is appreciated.

**Reply by informarte** [reply]
Workaround until the fix becomes available:
- Generate the IDEA project files.
- Manually add the language-level definition to `.idea/libraries/scala3_library_3_3_0_2_jar.xml`.
- Import the project into IDEA.

**Reply by lefou** [reply]
I updated the PR, should work now.

**Reply by informarte** [reply]
Yes, now it works.

**Comment by hmf** [comment]
@lefou 
Related question: I now work on VSCode + Metals to use Scala 3. It uses Bloop and seems to work ok (Java does not work well). If I understand correctly this means that metals uses Mill's contributed BSP support, but this *only* exports the project settings to Bloop, correct?

**Reply by lefou** [reply]
Metals is using Mills Bloop plugin (which exports config and fires bloop as BSP server) by default. This can be switched to Mill built-in BSP support. As far as I heard, Bloop works partially well (mostly for Scala modules), but is currently more robust than built-in BSP, which worked sometimes in the past better that Bloop. (I think we probably broke mill BSP since then.) I believe there is lots of easy to gain potential in built-in BSP support.

**Reply by lefou** [reply]
I reworked BSP support in Mill 0.10.0-M4. Give it a try. https://com-lihaoyi.github.io/mill/mill/Intro_to_Mill.html#_ide_support

**Reply by informarte** [reply]
I tried M4's BSP support together with IntelliJ IDEA 2021.2.2 on this project: https://github.com/informarte/yuck, branch `spikes/mill-0.10.0-M4`.

For the `main` part, everything seems to be ok, but there are issues with `test`. IDEA cannot navigate from test to production code (e.g. try to go to the definition of `Space` from line 21 of `AlldistinctTest.scala`) and it cannot run tests because it fails to find project-specific test infrastructure:
```
Exception in thread "main" java.lang.NoClassDefFoundError: yuck/util/logging/YuckLogging
```
(You can try any unit test, e.g. `AlldistinctTest`.)

It seems to me that there are classpath issues.

**Reply by lefou** [reply]
@informarte Thanks for testing and providing details. It may take some time until I can look into it.
